workload: interrupt worker loops when ramp period finishes

nvb opened the pull request:
This change fixes an issue that allowed tpcc to achieve over 100%
efficiency. This was possible because we weren't interrupting workers
when the load gen ramp period finished. This allowed long operations
like newOrder (which waits for over 18 seconds) to complete immediately
after the ramp period finished and we began recording stats, which
shouldn't have been possible.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27344)
<!-- Reviewable:end -->

danhhz reviewed (changes requested):
Trying to keep track of multiple contexts has been a huge source of bugs in backup/restore/import (this is why we replaced errgroup with ctxgroup). If we can, we should probably see if there's another way to do this.

An alternate idea is in `runRun`, ramp up the worker functions, then shut them all down, then bring them all back with no ramping. I don't have a ton of detail on the motivations for ramp, but would that accomplish the same thing?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-:-LH4olOG-IWQTVdvYMaw:b-xr4956)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/workload/run.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LH4p2FD2tPTh7L0PJCd:-LH4p2FD2tPTh7L0PJCe:bg6g2v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc4782f3a0f21a120a6db9e47070112166d7451/pkg/cmd/workload/run.go#L183)):*
> ```Go
> 
> 		if err := workFn(ctx); err != nil {
> 			if err != context.Canceled {
> ```

this also seems like a code smell


<!-- Sent from Reviewable.io -->

nvb committed:
workload: interrupt worker loops when ramp period finishes

This change fixes an issue that allowed tpcc to achieve over 100%
efficiency. This was possible because we weren't interrupting workers
when the load gen ramp period finished. This allowed long operations
like newOrder (which waits for over 18 seconds) to complete immediately
after the ramp period finished and we began recording stats, which
shouldn't have been possible.

Release note: None

nvb reviewed (commented):
Yeah, I think you're right. This was structured oddly. I re-worked it based on your suggestion so that the `workerRun` function no longer needs the strange context dance. It accomplishes the same goal but is a lot cleaner! PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-:-LI351x5793s999CG_o4:bs3xwq3)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/workload/run.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LH4p2FD2tPTh7L0PJCd:-LI35JeM4JiHu_IbUjAg:b-7r46nl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc4782f3a0f21a120a6db9e47070112166d7451/pkg/cmd/workload/run.go#L183)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this also seems like a code smell
</blockquote></details>

Ignoring context cancelation errors here doesn't seem to bad to me. What other alternatives are there?


<!-- Sent from Reviewable.io -->

danhhz reviewed (changes requested):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-:-LI7BnhcC1QmMVQ2J-wn:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/workload/run.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LH4p2FD2tPTh7L0PJCd:-LI7BMrPErpeemZtl954:bqmh97n) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc4782f3a0f21a120a6db9e47070112166d7451/pkg/cmd/workload/run.go#L183)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Ignoring context cancelation errors here doesn't seem to bad to me. What other alternatives are there?
</blockquote></details>

Feels less gross with the recent changes. Something with all the special casing and panics here still seems off but I think you'd have to switch from a waitgroup to a ctxgroup in the calling code to really address it and I don't want to piggyback too much on your good deed here.
___
*[pkg/cmd/workload/run.go, line 298 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LI7Apyw-BeEz6X_N9tF:-LI7Apyw-BeEz6X_N9tG:b-qy0m90) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e8d6b0c47638b02626dda8145ea6a14d461801e/pkg/cmd/workload/run.go#L298)):*
> ```Go
> 
> 			// Notify the process loop below to reset timers and histograms.
> 			close(rampDone)
> ```

Why not just inline that code here now? Seems more straightforward unless I'm missing something


<!-- Sent from Reviewable.io -->

danhhz reviewed (approved):
uhhhhh what happened reviewable?

nvb reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-:-LI7DUdk16PKI1ZZLqEp:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/workload/run.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LH4p2FD2tPTh7L0PJCd:-LI7CjC2D2g9EuSDiOjF:bq8zbzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc4782f3a0f21a120a6db9e47070112166d7451/pkg/cmd/workload/run.go#L183)):*
> Something with all the special casing and panics here still seems off but I think you'd have to switch from a waitgroup to a ctxgroup in the calling code to really address it

I'm curious about this. I guess I don't see how a ctxgroup would help here, but would certainly adjust this to use one if it would. Mind explaining? It seems to me that regardless of how this task is being managed, it still needs to handle cancellation internally, doesn't it?
___
*[pkg/cmd/workload/run.go, line 298 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LI7Apyw-BeEz6X_N9tF:-LI7CGC-7hkC1eun7SB0:bc78vmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e8d6b0c47638b02626dda8145ea6a14d461801e/pkg/cmd/workload/run.go#L298)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Why not just inline that code here now? Seems more straightforward unless I'm missing something
</blockquote></details>

Because it mutates a few variables that are scoped to the loop (`i` and `start`). This is easier and cleaner than needing to synchronize access to those variables.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27344)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/workload/run.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LH4p2FD2tPTh7L0PJCd:-LI7EFf05aJZDsUvuA-q:bpmo0zz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc4782f3a0f21a120a6db9e47070112166d7451/pkg/cmd/workload/run.go#L183)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> Something with all the special casing and panics here still seems off but I think you'd have to switch from a waitgroup to a ctxgroup in the calling code to really address it

I'm curious about this. I guess I don't see how a ctxgroup would help here, but would certainly adjust this to use one if it would. Mind explaining? It seems to me that regardless of how this task is being managed, it still needs to handle cancellation internally, doesn't it?
</blockquote></details>

Ugh, nevermind. I was hoping you could make this function ignorant of context cancelled errors entirely but the --tolerate-errors flag complicates that.

ctxgroup just means you can remove the panics
___
*[pkg/cmd/workload/run.go, line 298 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LI7Apyw-BeEz6X_N9tF:-LI7FTZ27_PhTDYmUEgh:b-q8ilnv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e8d6b0c47638b02626dda8145ea6a14d461801e/pkg/cmd/workload/run.go#L298)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Because it mutates a few variables that are scoped to the loop (`i` and `start`). This is easier and cleaner than needing to synchronize access to those variables.
</blockquote></details>

gotcha


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27344)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/workload/run.go, line 183 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27344#-LH4p2FD2tPTh7L0PJCd:-LI7JZH5BoqHH8R1xaIn:b2fdj0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cc4782f3a0f21a120a6db9e47070112166d7451/pkg/cmd/workload/run.go#L183)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Ugh, nevermind. I was hoping you could make this function ignorant of context cancelled errors entirely but the --tolerate-errors flag complicates that.

ctxgroup just means you can remove the panics
</blockquote></details>

Oh I see, because we could return it instead.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
bors r+
